internal/executor/thread-mate/service.go

```diff
@@ -0,0 +1,586 @@
+package thread_mate
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/gocarina/gocsv"
+	"github.com/google/uuid"
+	"github.com/olekukonko/tablewriter"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v3"
+
+	"github.com/kubeshop/botkube/internal/executor/x/mathx"
+	"github.com/kubeshop/botkube/internal/loggerx"
+	"github.com/kubeshop/botkube/pkg/api"
+	"github.com/kubeshop/botkube/pkg/api/executor"
+)
+
+const (
+	perPageItems     = 5
+	maxMsgContextLen = 64
+	ongoingCMName    = "ongoing-threads"
+	resolvedCMName   = "resolved-threads"
+)
+
+// ThreadMate represents the main component for managing threads and interactions.
+type ThreadMate struct {
+	log logrus.FieldLogger
+
+	systemData SystemData
+	assignees  []Assignee
+	membersLen uint32
+
+	resolvedThreads Threads
+	ongoingThreads  Threads
+
+	btnBuilder            *api.ButtonBuilder
+	cfgDumper             *ConfigMapDumper
+	cfg                   Config
+	lastProcessedActivity sync.Map
+}
+
+// New creates a new instance of ThreadMate.
+func New(cfg Config, cfgDumper *ConfigMapDumper) *ThreadMate {
+	var assignees []Assignee
+	for _, item := range cfg.RoundRobin.Assignees {
+		id, displayName, found := strings.Cut(item, ":")
+		if !found {
+			displayName = id
+		}
+		assignees = append(assignees, Assignee{ID: id, DisplayName: displayName})
+	}
+
+	return &ThreadMate{
+		log:       loggerx.New(cfg.Logger),
+		cfg:       cfg,
+		assignees: assignees,
+		systemData: SystemData{
+			roundRobin: RoundRobin{
+				//nolint:gosec // //nolint:gosec // false positive
+				next: uint32(rand.Int31n(int32(len(assignees)))), // randomize the first next person on each start
+			},
+		},
+		membersLen:            uint32(len(assignees)),
+		lastProcessedActivity: sync.Map{},
+		cfgDumper:             cfgDumper,
+		btnBuilder:            api.NewMessageButtonBuilder(),
+	}
+}
+
+// Start starts the ThreadMate component.
+func (t *ThreadMate) Start() {
+	t.ongoingThreads = t.tryToLoadOldThreads(ongoingCMName)
+	t.resolvedThreads = t.tryToLoadOldThreads(resolvedCMName)
+
+	go func() {
```

Comment:
shouldn't it take context and finish gracefully on `ctx.Done()`?

Reply:
we don't have context, I would need to do some signal channel and bind it to context. The one from the `Execute` method cannot be used as this needs to be a long-running operation, not per execute.

I think that we can do it later